eth/tracers/api.go

```diff
@@ -801,10 +801,13 @@ func containsTx(block *types.Block, hash common.Hash) bool {
 // TraceTransaction returns the structured logs created during the execution of EVM
 // and returns them as a JSON object.
 func (api *API) TraceTransaction(ctx context.Context, hash common.Hash, config *TraceConfig) (interface{}, error) {
-	_, blockHash, blockNumber, index, err := api.backend.GetTransaction(ctx, hash)
+	tx, blockHash, blockNumber, index, err := api.backend.GetTransaction(ctx, hash)
 	if err != nil {
 		return nil, err
 	}
+	if tx == nil {
+		return nil, nil
```

Comment:
Why not return the error `errTransactionNotFound` here instead of `nil`?

Reply:
Changing this to 

```suggestion
		return nil, errors.New("transaction not found")
```
We get 
```
> debug.traceTransaction("0x0000000000000000000000000000000000000000000000000000000000000000")
WARN [11-18|10:38:20.230] Served debug_traceTransaction            reqid=9 duration="127.204µs" err="transaction not found"
Error: transaction not found
	at web3.js:6365:9(45)
	at send (web3.js:5099:62(34))
	at <eval>:1:23(4)

```

Reply:
It can be written _slightly_ more compact using a switch 
```golang
	switch {
	case err != nil:
		return nil, err
	case tx == nil:
		return nil, errors.New("transaction not found")
	case blockNumber == 0:
		return nil, errors.New("genesis is not traceable")
	}
```

Reply:
> Why not return the error errTransactionNotFound here instead of nil?

I came here to say the same

Reply:
We have an existing convention in other RPC APIs returning `null` when the requested item is not found.

Reply:
That sounds fine to me too; I was trying to match the behavior of `eth_getTransactionByHash` and `eth_getTransactionReceipt` which both return `nil` for a not-found transaction. But happy to take that suggestion if you think it's better.

Reply:
If we want to return `errTransactionNotFound`, we could also do this in the `GetTransaction` method: https://github.com/ethereum/go-ethereum/blob/c2e0abce2eedc1ba2a1b32c46fd07ef18a25354a/eth/api_backend.go#L267-L270

Reply:
> We have an existing convention in other RPC APIs returning null when the requested item is not found.

@fjl so you think the pr as is (return `nil` rather than error) is preferrable? 

I don't really understand the motivation behind that, personally.

Reply:
The motivation for the behavior of `getTransactionByHash` and `getTransactionReceipt` is that 'not found' is not an error. You could imagine that, with an RPC client that throws an exception for any error response, the case where the requested item is not found should not raise an exception.

Another reason might be that the people who designed these APIs didn't know that JSON-RPC has error codes. Returning `null` provides a way to recognize this specific condition. 

Returning `errors.New("transaction not found")` here is not great because it will not return any specific error code, and callers will not be able to know whether the error is related to a database/disk fault or simply means the item is not there.

Reply:
FWIW it's already returning an error: `genesis is not traceable`. So returning `transaction not found` would retain the existing behavior, with an error message that makes more sense.